packages/medusa/src/services/payment-collection.ts

```diff
@@ -290,10 +323,114 @@ export default class PaymentCollectionService extends TransactionBaseService {
     })
   }
 
+  /**
+   * Manages a single payment sessions of a payment collection.
+   * @param paymentCollectionId - the id of the payment collection
+   * @param sessionsInput - object containing payment session info
+   * @param customerId - the id of the customer
+   * @return the payment collection and its payment session.
+   */
+  async setPaymentSession(
+    paymentCollectionId: string,
+    sessionInput: PaymentCollectionsSessionsInput,
+    customerId: string
+  ): Promise<PaymentCollection> {
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const paymentCollectionRepository = manager.getCustomRepository(
+        this.paymentCollectionRepository_
+      )
+
+      const payCol = await this.retrieve(paymentCollectionId, {
+        relations: ["region", "region.payment_providers", "payment_sessions"],
+      })
+
+      if (payCol.status !== PaymentCollectionStatus.NOT_PAID) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `Cannot set payment sessions for a payment collection with status ${payCol.status}`
+        )
+      }
+
+      const hasProvider = payCol?.region?.payment_providers
+        .map((p) => p.id)
+        .includes(sessionInput.provider_id)
+
+      if (!hasProvider) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "Payment provider not found"
+        )
+      }
+
+      const customer = await this.customerService_
+        .withTransaction(manager)
+        .retrieve(customerId, {
+          select: ["id", "email", "metadata"],
+        })
+        .catch(() => null)
+
+      const paymentSessions: PaymentSession[] = []
+      const inputData: PaymentProviderDataInput = {
+        resource_id: payCol.id,
+        currency_code: payCol.currency_code,
+        amount: payCol.amount,
+        provider_id: sessionInput.provider_id,
+        customer,
```

Comment:
thought: In the Stripe payment plugin, we have the following check when updating a payment session:
```
      const stripeId = paymentInput.customer?.metadata?.stripe_id

      if (stripeId !== paymentInput.customer_id) {
        return await this.createPaymentNew(paymentInput, intentRequestData)
      } else {
        if (paymentSessionData.amount === Math.round(paymentInput.amount)) {
          return paymentSessionData
        }

        return await this.stripe_.paymentIntents.update(paymentSessionData.id, {
          amount: Math.round(paymentInput.amount),
        })
      }
```

But with this new implementation, I don't think we ever add `customer_id` in the flow from the payment collection service to the payment provider plugin service. Shouldn't we add that alongside the customer to make the logic in the Stripe provider kick in properly?

Reply:
Thanks for seeing this. That actually uncovered a bigger issue when the customer is empty. customerService.retrieve is fetching a customer when the customerId is undefined and currently stripe plugin is requiring at least a customer email to work.
I will address these issues.

Reply:
For that PR I'm checking the customerId undefined before calling customerService.retrive, but I will try to handle these undefined ids in all retrieve methods in another PR.
This can cause some weird hidden issues in multiple places

Reply:
Great - sounds good!